overwatch-rs/src/overwatch/life_cycle.rs

```diff
@@ -0,0 +1,49 @@
+use crate::services::life_cycle::{FinishedSignal, LifecycleHandle, LifecycleMessage};
+use crate::services::ServiceId;
+use crate::DynError;
+use std::collections::HashMap;
+use tokio::sync::broadcast::Sender;
+
+#[derive(Clone)]
+pub struct ServicesLifeCycleHandle {
+    handlers: HashMap<ServiceId, LifecycleHandle>,
+}
+
+impl ServicesLifeCycleHandle {
+    pub fn shutdown(
+        &self,
+        service: ServiceId,
+        sender: Sender<FinishedSignal>,
+    ) -> Result<(), DynError> {
+        self.handlers
+            .get(service)
+            .unwrap()
+            .send(LifecycleMessage::Shutdown(sender))?;
+        Ok(())
+    }
+
+    pub fn kill(&self, service: ServiceId) -> Result<(), DynError> {
+        self.handlers
+            .get(service)
+            .unwrap()
+            .send(LifecycleMessage::Kill)
+    }
+    pub fn kill_all(&self) -> Result<(), DynError> {
+        for service_id in self.services_ids() {
+            self.kill(service_id)?;
+        }
+        Ok(())
+    }
+
+    pub fn services_ids(&self) -> impl Iterator<Item = ServiceId> + '_ {
+        self.handlers.keys().copied()
+    }
+}
+
+impl<I: IntoIterator<Item = (ServiceId, LifecycleHandle)>> From<I> for ServicesLifeCycleHandle {
```

Comment:
Nit: Even though we have const checks for duplicated `ServiceId`, will it be better to implement `TryFrom` for `ServiceLifeCycleHandle`? What is your opinion?

Reply:
What do you mean exactly? Substituting the `(ServiceId, LifecycleHandle)` for a `TryFrom<LifecycleHandle>`? Problem with that is that the handle is not bound to a Service itself, otherwise would be difficult to fit the types unless making a trait for the handle and boxing them. So we have no access to the `ServiceId` otherwise. Anyway, in this case the `ServiceId` is extracted from the deriving implementation itself.
Or you meant something else? Go ahead and make some specific proposition example 😃

Reply:
Yeah, what I mean is change to

```rust
impl<I: IntoIterator<Item = (ServiceId, LifecycleHandle)>> TryFrom<I> for ServicesLifeCycleHandle {}
```

My concerns are a `From` behavior here without handling duplicated `ServiceId` is a little bit invalidates what we do in derive macro. As `ServicesLifeCycleHandle` will be exported to the end user, and we cannot make sure all of end users will use `derive(Services)`, they may manually implement the `Services` trait.

Reply:
Nice argument!, makes sense. I'll add some check for that and change into `TryFrom`.